File: 2-ad-hoc/1206.cpp
#include<iostream>
//
// °~°
//
int**pieces,King,**board;
void translate(std::string&gross,int index){
	pieces[index][0]=(gross[0]=='P')?1:(gross[0]=='B')?2:(gross[0]=='T')?3:(gross[0]=='W')?0:4;
	pieces[index][1]=int(gross[1])-97+(gross[2]-'0'-1)*10;
	board[(pieces[index][1]-pieces[index][1]%10)/10][pieces[index][1]%10]=2;
}

void intersectionHorizont(int yx){
	int x=yx%10,y=(yx-x)/10;
	for(int auxX=x,auxY=y+1;auxY<8;++auxY){
		if(board[auxY][auxX]>1){
			++board[auxY][auxX];
			break;
		}
		board[auxY][auxX]=1;
	}
	for(int auxX=x,auxY=y-1;auxY>-1;--auxY){
		if(board[auxY][auxX]>1){
			++board[auxY][auxX];
			break;
		}
		board[auxY][auxX]=1;
	}
	for(int auxX=x+1,auxY=y;auxX<8;++auxX){
		if(board[auxY][auxX]>1){
			++board[auxY][auxX];
			break;
		}
		board[auxY][auxX]=1;
	}
	for(int auxX=x-1,auxY=y;auxX>-1;--auxX){
		if(board[auxY][auxX]>1){
			++board[auxY][auxX];
			break;
		}
		board[auxY][auxX]=1;
	}
}
void intersectionVertical(int yx){
	int x=yx%10,y=(yx-x)/10;
	for(int auxX=x-1,auxY=y-1;auxX>-1&&auxY>-1;--auxX,--auxY){
		if(board[auxY][auxX]>1){
			++board[auxY][auxX];
			break;
		}
		board[auxY][auxX]=1;
	}
	for(int auxX=x+1,auxY=y-1;auxX<8&&auxY>-1;++auxX,--auxY){
		if(board[auxY][auxX]>1){
			++board[auxY][auxX];
			break;
		}
		board[auxY][auxX]=1;
	}
	for(int auxX=x+1,auxY=y+1;auxX<8&&auxY<8;++auxX,++auxY){
		if(board[auxY][auxX]>1){
			++board[auxY][auxX];
			break;
		}
		board[auxY][auxX]=1;
	}
	for(int auxX=x-1,auxY=y+1;auxX>-1&&auxY<8;--auxX,++auxY){
		if(board[auxY][auxX]>1){
			++board[auxY][auxX];
			break;
		}
		board[auxY][auxX]=1;
	}
}
void intersectionPawn(int yx){
	int x=yx%10,y=(yx-x)/10;
	if(x-1>-1&&y-1>-1)board[y-1][x-1]=1;
	if(x+1<8&&y-1>-1)board[y-1][x+1]=1;
}
void intersectionKing(int yx){
	int x=yx%10,y=(yx-x)/10;
	if(x+1<8)board[y][x+1]=1;
	if(x-1>-1)board[y][x-1]=1;
	for(int auxX=x-1;auxX<8&&auxX<x+2&&y+1<8;++auxX)board[y+1][auxX]=1;
	for(int auxX=x-1;auxX<8&&auxX<x+2&&y-1>-1;++auxX)board[y-1][auxX]=1;
}


int notCheckmate(int yx,int size){
	for(int i=0;i<size;++i)board[(pieces[i][1]-pieces[i][1]%10)/10][pieces[i][1]%10]-=2;
	int x=yx%10,y=(yx-x)/10;
	if((!board[y][x-1]&&x-1>-1)|(!board[y][x+1]&&x+1<8)|!board[y][x])return 1;
	for(int auxX=(x-1>-1)?x-1:x;auxX<8&&auxX<x+2&&y+1<8;++auxX)if(!board[y+1][auxX])return 1;
	for(int auxX=(x-1>-1)?x-1:x;auxX<8&&auxX<x+2&&y-1>-1;++auxX)if(!board[y-1][auxX])return 1;
	return 0;
}
void inCheck(int amountPieces){
	for(int i=0;i<amountPieces;++i){
		//
		//std::cout<<pieces[i][0]<<"|";
		if(pieces[i][0]==0)intersectionKing(pieces[i][1]);
		if(pieces[i][0]==1)intersectionPawn(pieces[i][1]);
		if(pieces[i][0]==2|pieces[i][0]==4)intersectionVertical(pieces[i][1]);
		if(pieces[i][0]==3|pieces[i][0]==4)intersectionHorizont(pieces[i][1]);
		//
	}
}
//
int main(){
	int rep;
	while(std::cin>>rep){
		board=(int**)malloc(sizeof(int*)*8);
		for(int i=0;i<8;++i)board[i]=(int*)calloc(8,sizeof(int));
		pieces=(int**)malloc(sizeof(int*)*rep);
		for(int i=0;i<rep;++i){
			std::string temp;
			std::cin>>temp;
			pieces[i]=(int*)malloc(sizeof(int)*2);
			translate(temp,i);
		}
		inCheck(rep);
		/*
		std::cout<<"\n";
		for(int i=7;i>-1;--i){
			for(int j=0;j<8;++j)std::cout<<board[i][j]<<" ";
			std::cout<<"\n";
		}*/	
		std::string init;
		std::cin>>init;
		King=int(init[1])-97+(init[2]-'0'-1)*10;
		if(!notCheckmate(King,rep))std::cout<<"SIM\n";
		else std::cout<<"NAO\n";
		free(board);
	}
	return 0;
}

File: 2-ad-hoc/1206.py
board,data=None,None

def Translate(string):
    data.append([None,None])
    data[-1][0]=0 if string[0]=='W' else 1 if string[0]=='P' else 2 if string[0]=='T' else 3 if string[0]=='B' else 4
    data[-1][1]=(ord(string[1])-97)*10+ord(string[2])-ord('0')-1
    board[data[-1][1]%10][data[-1][1]//10]=2

def intersectionHorizontal(xy):
    x,y=xy//10,xy%10-1
    while  y>-1:
        if board[y][x]>1:
            board[y][x]+=1
            break
        board[y][x]=1
        y-=1
    y=xy%10+1
    while y<8:
        if board[y][x]>1:
            board[y][x]+=1
            break
        board[y][x]=1
        y+=1
    x-=1
    y=xy%10
    while x>-1:
        if board[y][x]>1:
            board[y][x]+=1
            break
        board[y][x]=1
        x-=1
    x=xy//10+1
    while x<8:
        if board[y][x]>1:
            board[y][x]+=1
            break
        board[y][x]=1
        x+=1
def intersectionVertical(xy):
    x,y=xy//10-1,xy%10-1
    while x>-1 and y>-1:
        if board[y][x]>1:
            board[y][x]+=1
            break
        board[y][x]=1
        x-=1
        y-=1
    x,y=xy//10+1,xy%10-1
    while x<8 and y>-1:
        if board[y][x]>1:
            board[y][x]+=1
            break
        board[y][x]=1
        x+=1
        y-=1
    x,y=xy//10-1,xy%10+1
    while x>-1 and y<8:
        if board[y][x]>1:
            board[y][x]+=1
            break
        board[y][x]=1
        x-=1
        y+=1
    x,y=xy//10+1,xy%10+1
    while x<8 and y<8:
        if board[y][x]>1:
            board[y][x]+=1
            break
        board[y][x]=1
        x+=1
        y+=1
def intersectionKing(xy):
    for i in range(3):
        x,y=xy//10-1,xy%10-1+i
        if y<0 or y>7:continue
        for j in range(3):
            if x+j<0 or x+j>7 or (x+j)*10+y==xy:continue
            if board[y][x+j]>1:board[y][x+j]+=1
            else: board[y][x+j]=1
def intersectionPawn(xy):
    x,y=xy//10,xy%10-1
    if y<0:return;
    if x+1<8:
        if board[y][x+1]>1:board[y][x+1]+=1
        else: board[y][x+1]=1
    if x-1>-1:
        if board[y][x-1]>1:board[y][x-1]+=1
        else: board[y][x-1]=1
def intersections():
    for i in data:
        if i[0]==0:intersectionKing(i[1])
        elif i[0]==1: intersectionPawn(i[1])
        elif i[0]==2 or i[0]==4: intersectionHorizontal(i[1])
        if i[0]==3 or i[0]==4:intersectionVertical(i[1])
    for i in data:
        board[i[1]%10][i[1]//10]-=2
def inCheck(xy):
    for i in range(3):
        x,y=xy//10-1,xy%10-1+i
        if y<0 or y>7:continue
        for j in range(3):
            if x+j<0 or x+j>7:continue
            if not board[y][x+j]:return 0
    return 1
while 1:
    try:
        input()
        data=[]
        board=[[0 for i in range(8)]for j in range(8)]
        for i in input().split():Translate(i)
        intersections()
        King=input()
        print('NAO') if not inCheck((ord(King[1])-97)*10+ord(King[2])-ord('0')-1) else print('SIM')
        board,data=None,None
    except EOFError:break
